doc/clust.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "CLUST" "1" "September 2017" "www.complex-networks.net" "www.complex-networks.net"
.
.SH "NAME"
\fBclust\fR \- Compute the graph and node clustering coefficients
.
.SH "SYNOPSIS"
\fBclust\fR \fIgraph_in\fR [SHOW]
.
.SH "DESCRIPTION"
\fBclust\fR computes the clustering coefficient of the undirected graph given as input in the file \fIgraph_in\fR\. If \fBSHOW\fR is provided as a second parameter, the program prints on STDERR the label, degree, and clustering coefficient of all the nodes in \fIgraph_in\fR\.
.
.SH "PARAMETERS"
.
.TP
\fIgraph_in\fR
undirected input graph (edge list)\. If is equal to \fB\-\fR (dash), read the edge list from STDIN\.
.
.TP
SHOW
If the second (optional) parameter is equal to \fBSHOW\fR, the program will dump on the standard error the label, degree, and clustering coefficient of each node in \fIgraph_in\fR\.
.
.SH "OUTPUT"
If only \fIgraph_in\fR is specified, then the output is a single line, containing the clustering coefficient of the undirected graph provided as input\. If \fBSHOW\fR is specified, the program will print on the standard output one line for each node, in the format:
.
.IP "" 4
.
.nf

node_1 k_1 c_1
node_2 k_2 c_2
node_3 k_3 c_3
 \.\.\.\.
.
.fi
.
.IP "" 0
.
.P
where \fBnode_1\fR is the label of the node, \fBk_1\fR is its degree, and \fBc_1\fR is its node clustering coefficient\.
.
.SH "EXAMPLES"
The most simple way of using \fBclust\fR is to compute only the clustering coefficient of a graph\. For instance:
.
.IP "" 4
.
.nf

      $ clust er_1000_5000\.txt
      0\.01034196
      $
.
.fi
.
.IP "" 0
.
.P
will show on output the clustering coefficient of the graph \fBer_1000_5000\.txt\fR\. In order to obtain the clustering coefficient of all the nodes, we should use:
.
.IP "" 4
.
.nf

      $ clust er_1000_5000\.txt SHOW
      0 10 0\.0222222
      1 3 0
      2 7 0
      3 5 0
      4 10 0
      5 17 0
      6 14 0
      7 8 0
      8 6 0
      9 11 0
      10 9 0
      11 10 0
      12 13 0\.0128205
       \.\.\.\.
      998 10 0\.0222222
      999 9 0
      0\.01034196
      $
.
.fi
.
.IP "" 0
.
.P
The last line printed on output is still the value of the clustering coefficient of the graph, while the previous 1000 lines (which are printed on STDERR) contain the label, degree, and clustering coefficient of all the nodes\. For instance, the first line indicates that node \fB0\fR has degree equal to \fB10\fR and clustering coefficient equal to \fB0\.0222222\fR\. It is more convenient to save the values of node clustering coefficients in a file, e\.g\.:
.
.IP "" 4
.
.nf

      $ clust er_1000_5000\.txt SHOW  2> er_1000_5000\.txt_node_clust
      0\.01034196
      $
.
.fi
.
.IP "" 0
.
.P
In this case, the program prints on output only the graph clustering coefficient \fB0\.01034196\fR, while the node clustering coefficients are saved on the file \fBer_1000_5000\.txt_node_clust\fR (notice the syntax \fB2> er_1000_5000\.txt_node_clust\fR, which redirects the STDERR to the file \fBer_1000_5000\.txt_node_clust\fR)\.
.
.SH "SEE ALSO"
clust_w(1)
.
.SH "REFERENCES"
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Chapter 4, Cambridge University Press (2017)
.
.IP "" 0
.
.SH "AUTHORS"
(c) Vincenzo \'KatolaZ\' Nicosia 2009\-2017 \fB<v\.nicosia@qmul\.ac\.uk>\fR\.


debian/changelog
netbunch (1.0-1) unstable; urgency=medium

  * Initial release 

 -- Vincenzo (KatolaZ) Nicosia <katolaz@freaknet.org>  Thu, 28 Sep 2017 06:46:47 +0100


src/clust/clust.c
/**
 *  This program is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 *  (c) Vincenzo Nicosia 2009-2017 -- <v.nicosia@qmul.ac.uk>
 * 
 *  This file is part of NetBunch, a package for complex network
 *  analysis and modelling. For more information please visit:
 *
 *             http://www.complex-networks.net/
 *
 *  If you use this software, please add a reference to 
 *
 *               V. Latora, V. Nicosia, G. Russo             
 *   "Complex Networks: Principles, Methods and Applications"
 *              Cambridge University Press (2017) 
 *                ISBN: 9781107103184
 *
 ***********************************************************************
 *
 *  This program computes the clustering coefficient of a graph and of
 *  each node.
 *
 *
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "utils.h"


void usage(char *argv[]){

    printf("********************************************************************\n"
         "**                                                                **\n"
         "**                        -*-    clust   -*-                      **\n"
         "**                                                                **\n"
         "**   Compute the graph clustering coefficient and, optionally,    **\n"
         "**   also the node clustering coefficient of all the nodes in     **\n"
         "**   the graph 'graph_in'.                                        **\n"
         "**                                                                **\n"
         "**   The input file 'graph_in' is an edge-list:                   **\n"
         "**                                                                **\n"
         "**                            I_1 J_1                             **\n"
         "**                            I_2 J_2                             **\n"
         "**                            I_3 J_3                             **\n"
         "**                            ... ...                             **\n"
         "**                            I_K J_K                             **\n"
         "**                                                                **\n"
         "**    If 'graph_in' is equal to '-' (dash), read the file from    **\n" 
         "**    the standard input (STDIN).                                 **\n"
         "**                                                                **\n"
         "**    The output is a single line containing the value of the     **\n"
         "**    graph clustering coefficient of 'graph_in'.                 **\n"
         "**                                                                **\n"
         "**    If 'SHOW' is provided as second argument, the program       **\n"
         "**    on the standard error (STDERR) the label, degree, and       **\n"
         "**    node clustering coefficient of all the nodes of 'graph_in'  **\n"
         "**    in the format:                                              **\n"
         "**                                                                **\n"
         "**      node_1 k_1 c_1                                            **\n"
         "**      node_2 k_2 c_2                                            **\n"
         "**      node_3 k_3 c_3                                            **\n"
         "**      node_4 k_4 c_4                                            **\n"
         "**      .....                                                     **\n"
         "**                                                                **\n"
         "**    where 'k_1' is the degree of 'node_1' and 'c_1' is its      **\n"
         "**    node clustering coefficient. If a node has degree smaller   **\n"
         "**    than 2, its clustering coefficient is equal to zero.        **\n"
         "**                                                                **\n"
         "********************************************************************\n"
         " This is Free Software - You can use and distribute it under \n"
         " the terms of the GNU General Public License, version 3 or later\n\n"
         " Please visit http://www.complex-networks.net for more information\n\n"
         " (c) Vincenzo Nicosia 2009-2017 (v.nicosia@qmul.ac.uk)\n"
         "********************************************************************\n\n"
         );
  
  printf("Usage: %s <graph_in> [SHOW]\n", argv[0]);
}



double avg_clust(unsigned int *J_slap, unsigned int *r_slap, unsigned int N, int show){

  double c = 0, ci=0;
  unsigned int i, j, l, ki;

  for(i=0; i<N; i ++){/* for each node i..*/
    ci = 0;
    ki = r_slap[i+1] - r_slap[i];
    for(j=r_slap[i]; j<r_slap[i+1]; j++){/*... pick a neighbour...*/
      for (l=r_slap[J_slap[j]]; l<r_slap[J_slap[j]+1]; l++){
        if (J_slap[l] != i){
          /*...and check how many times it has a link to a neighbour of i */
          if(is_neigh(J_slap, r_slap, N, i, J_slap[l])){
            ci += 1;
          }
        }
      }
    }
    if(ki > 1)
      ci /= ki * (ki-1);
    else
      ci = 0;
    if (show) 
      fprintf(stderr, "%d %d %g\n", i, ki, ci);
    c += ci;
  }
  c /= N;
  return c;
}




int main(int argc, char *argv[]){

  unsigned int *J_slap=NULL, *r_slap=NULL;
  unsigned int K, N;
  FILE *filein;
  double c;
  int show = 0;

  if (argc < 2){
    usage(argv);
    exit(1);
  }
  
  if (argc >2 && !my_strcasecmp("SHOW", argv[2])){
    show = 1;
  }

  if (!strcmp(argv[1], "-")){
    /* take the input from STDIN */
    filein = stdin;
  }
  else {
    filein = openfile_or_exit(argv[1], "r", 2);
  }

  read_slap(filein, &K, &N, &J_slap, &r_slap);
  fclose(filein);
  
  c = avg_clust(J_slap, r_slap, N, show);
  printf("%2.8lf\n", c);
  free(J_slap);
  free(r_slap);
}
